x/evm/types/config.go

```diff
@@ -0,0 +1,152 @@
+	istanbulBlock := sdk.ZeroInt()
+	muirGlacierBlock := sdk.ZeroInt()
+	berlinBlock := sdk.ZeroInt()
+	londonBlock := sdk.ZeroInt()
+	arrowGlacierBlock := sdk.ZeroInt()
+	grayGlacierBlock := sdk.ZeroInt()
+	mergeNetsplitBlock := sdk.ZeroInt()
+
+	return ChainConfig{
+		HomesteadBlock:      &homesteadBlock,
+		DAOForkBlock:        &daoForkBlock,
+		DAOForkSupport:      true,
+		EIP150Block:         &eip150Block,
+		EIP155Block:         &eip155Block,
+		EIP158Block:         &eip158Block,
+		ByzantiumBlock:      &byzantiumBlock,
+		ConstantinopleBlock: &constantinopleBlock,
+		PetersburgBlock:     &petersburgBlock,
+		IstanbulBlock:       &istanbulBlock,
+		MuirGlacierBlock:    &muirGlacierBlock,
+		BerlinBlock:         &berlinBlock,
+		LondonBlock:         &londonBlock,
+		ArrowGlacierBlock:   &arrowGlacierBlock,
+		GrayGlacierBlock:    &grayGlacierBlock,
+		MergeNetsplitBlock:  &mergeNetsplitBlock,
+		ShanghaiTime:        0,
+		CancunTime:          0,
+		PragueTime:          0,
+		VerkleTime:          0,
+	}
+}
+
+func getBlockValue(block *sdk.Int) *big.Int {
+	if block == nil || block.IsNegative() {
+		return nil
+	}
+
+	return block.BigInt()
+}
+
+func (cc ChainConfig) Validate() error {
+	if err := validateBlock(cc.HomesteadBlock); err != nil {
+		return errors.New("homesteadBlock")
+	}
+	if err := validateBlock(cc.DAOForkBlock); err != nil {
+		return errors.New("daoForkBlock")
+	}
+	if err := validateBlock(cc.EIP150Block); err != nil {
+		return errors.New("eip150Block")
+	}
+	if err := validateBlock(cc.EIP155Block); err != nil {
+		return errors.New("eip155Block")
+	}
+	if err := validateBlock(cc.EIP158Block); err != nil {
+		return errors.New("eip158Block")
+	}
+	if err := validateBlock(cc.ByzantiumBlock); err != nil {
+		return errors.New("byzantiumBlock")
+	}
+	if err := validateBlock(cc.ConstantinopleBlock); err != nil {
+		return errors.New("constantinopleBlock")
+	}
+	if err := validateBlock(cc.PetersburgBlock); err != nil {
+		return errors.New("petersburgBlock")
+	}
+	if err := validateBlock(cc.IstanbulBlock); err != nil {
+		return errors.New("istanbulBlock")
+	}
+	if err := validateBlock(cc.MuirGlacierBlock); err != nil {
+		return errors.New("muirGlacierBlock")
+	}
+	if err := validateBlock(cc.BerlinBlock); err != nil {
+		return errors.New("berlinBlock")
+	}
+	if err := validateBlock(cc.LondonBlock); err != nil {
+		return errors.New("londonBlock")
+	}
+	if err := validateBlock(cc.ArrowGlacierBlock); err != nil {
+		return errors.New("arrowGlacierBlock")
+	}
+	if err := validateBlock(cc.GrayGlacierBlock); err != nil {
+		return errors.New("GrayGlacierBlock")
+	}
+	if err := validateBlock(cc.MergeNetsplitBlock); err != nil {
+		return errors.New("MergeNetsplitBlock")
+	}
+	if err := cc.EthereumConfig(nil).CheckConfigForkOrder(); err != nil {
+		return errors.New("invalid config fork order")
+	}
+	return nil
+}
+
+func validateBlock(block *sdk.Int) error {
+	if block == nil {
+		return nil
+	}
+
+	if block.IsNegative() {
```

Comment:
nit block > 0?

Reply:
`block` can be 0 in the config which would indicate that the corresponding upgrade was ineligible (i.e. the chain started with the new version). I'll add comments in this file to better explain it